core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1663,11 +1660,35 @@ object Stream {
           }
           only match {
             case None =>
-              hd.map(f).foldRightLazy(Stream.fromFreeC(tl).flatMap(f))(_ ++ _).get
-            case Some(o) => f(o).get
+              // tail mail contain algebra that is critical for proper sequenced scope closure
+              // as such here in the event of the failure we will always drain the algebra
+              // for any `CloseScope` so we will close correct resources and set correct parent scope.
+              // note that resources are guaranteed to be closed by the way how Scope.close is defined, however
+              // in that case, in certain situations, the flatMap followed by `handleErrorWith` may
+              // without this change postpone closure of resources until the handler in handleErrorWith is interpreted.
+
+              def fby( s1: Either[O, Stream[F, O]], s2: => Stream[F,O2]): Stream[F, O2] = {
```

Comment:
I'm not convinced we should ever be searching the tail for more work -- seems like this is patching for something else that's wrong.

Reply:
yes, what wrong is that we skip all cleanups on error. Thats it. And in that case we rely on the cleanups to be run by `Algebra.runFold`.

Reply:
If we would signal interrupt by exception, that's likely only way. We could perhaps decide it not to signal by exception but somehow build it in the FreeC itself, but I do not have a good solution for that either.